File: 141_linked_list_cycle/linked_list_cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode s = head, q = head;
        while (s != null && q != null) {
            s = s.next;
            if (q.next == null) {
                break;
            }
            q = q.next.next;
            if (s == q) {
                return true;
            }
        }
        return false;
    }
}

File: 141_linked_list_cycle/linked_list_cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    struct ListNode * s = head, * q = head;
    while (s != NULL && q != NULL) {
        s = s->next;
        if (q->next == NULL) {
            break;
        }
        q = q->next->next;
        if (s == q) {
            return true;
        }
    }
    return false;
}
